File: ios/Nimbus/Services/PhotoService.swift
import CoreLocation
import Foundation

/// The shared half of Nimbus: leaving photos at a place and finding the ones
/// other people left.
///
/// Kept as a protocol so the local Node server can be swapped for a hosted
/// backend (Supabase, Firebase, anything) without touching a single view.
protocol PhotoService {
    func health() async -> Bool

    /// Announce this device's identity and get back its friend code and current
    /// friends. Safe to call on every launch; it doubles as a rename.
    func register(_ explorer: Explorer) async throws -> RegistrationResponse
    func friends(of viewerID: String) async throws -> [RemoteUser]
    func addFriend(code: String, for viewerID: String) async throws -> AddFriendResponse

    func nearby(coordinate: CLLocationCoordinate2D, viewerID: String) async throws -> NearbyResult
    func photos(inLatitudes latitudes: ClosedRange<Double>,
                longitudes: ClosedRange<Double>,
                viewerID: String) async throws -> [Photo]
    func upload(imageData: Data,
                coordinate: CLLocationCoordinate2D,
                caption: String,
                explorer: Explorer,
                placeName: String?) async throws -> UploadResponse
}

enum PhotoServiceError: LocalizedError {
    case server(status: Int, message: String)
    case unreachable

    var errorDescription: String? {
        switch self {
        case .server(let status, let message): "Server said \(status): \(message)"
        case .unreachable: "Can't reach the Nimbus server. Is `node server.js` running?"
        }
    }
}

/// Talks to server/server.js.
final class NimbusAPI: PhotoService {

    private let baseURL: URL
    private let session: URLSession

    init(baseURL: URL = Config.serverBaseURL) {
        self.baseURL = baseURL

        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = 8
        // Seed images are served immutable, so the cache does the heavy lifting
        // when the same photos come back from several searches.
        configuration.requestCachePolicy = .useProtocolCachePolicy
        configuration.urlCache = URLCache(memoryCapacity: 32 << 20, diskCapacity: 256 << 20)
        self.session = URLSession(configuration: configuration)
    }

    func health() async -> Bool {
        do {
            let (_, response) = try await session.data(from: baseURL.appendingPathComponent("health"))
            return (response as? HTTPURLResponse)?.statusCode == 200
        } catch {
            return false
        }
    }

    func register(_ explorer: Explorer) async throws -> RegistrationResponse {
        try await post(
            "users",
            body: [
                "id": explorer.id,
                "displayName": explorer.displayName,
                "color": explorer.colorHex,
            ]
        )
    }

    func friends(of viewerID: String) async throws -> [RemoteUser] {
        var components = URLComponents(
            url: baseURL.appendingPathComponent("friends"),
            resolvingAgainstBaseURL: false
        )!
        components.queryItems = [URLQueryItem(name: "userId", value: viewerID)]
        let response: FriendListResponse = try await get(components.url!)
        return response.friends
    }

    func addFriend(code: String, for viewerID: String) async throws -> AddFriendResponse {
        try await post("friends", body: ["userId": viewerID, "code": code])
    }

    func nearby(coordinate: CLLocationCoordinate2D, viewerID: String) async throws -> NearbyResult {
        var components = URLComponents(
            url: baseURL.appendingPathComponent("photos/nearby"),
            resolvingAgainstBaseURL: false
        )!
        components.queryItems = [
            URLQueryItem(name: "lat", value: String(coordinate.latitude)),
            URLQueryItem(name: "lon", value: String(coordinate.longitude)),
            URLQueryItem(name: "viewerId", value: viewerID),
        ]
        return try await get(components.url!)
    }

    func photos(inLatitudes latitudes: ClosedRange<Double>,
                longitudes: ClosedRange<Double>,
                viewerID: String) async throws -> [Photo] {
        var components = URLComponents(
            url: baseURL.appendingPathComponent("photos/bbox"),
            resolvingAgainstBaseURL: false
        )!
        components.queryItems = [
            URLQueryItem(name: "minLat", value: String(latitudes.lowerBound)),
            URLQueryItem(name: "maxLat", value: String(latitudes.upperBound)),
            URLQueryItem(name: "minLon", value: String(longitudes.lowerBound)),
            URLQueryItem(name: "maxLon", value: String(longitudes.upperBound)),
            // Without this the server would hand back everybody's photographs.
            URLQueryItem(name: "viewerId", value: viewerID),
        ]
        let response: PhotoListResponse = try await get(components.url!)
        return response.photos
    }

    func upload(imageData: Data,
                coordinate: CLLocationCoordinate2D,
                caption: String,
                explorer: Explorer,
                placeName: String?) async throws -> UploadResponse {
        var request = URLRequest(url: baseURL.appendingPathComponent("photos"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 20

        let body: [String: Any] = [
            "userId": explorer.id,
            "displayName": explorer.displayName,
            "color": explorer.colorHex,
            "lat": coordinate.latitude,
            "lon": coordinate.longitude,
            "caption": caption,
            "takenAt": Date().timeIntervalSince1970 * 1000,
            "placeName": placeName as Any,
            "imageBase64": imageData.base64EncodedString(),
        ].compactMapValues { $0 is NSNull ? nil : $0 }

        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        return try await send(request)
    }

    // MARK: Transport

    private func get<T: Decodable>(_ url: URL) async throws -> T {
        try await send(URLRequest(url: url))
    }

    private func post<T: Decodable>(_ path: String, body: [String: Any]) async throws -> T {
        var request = URLRequest(url: baseURL.appendingPathComponent(path))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        return try await send(request)
    }

    private func send<T: Decodable>(_ request: URLRequest) async throws -> T {
        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw PhotoServiceError.unreachable
        }

        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(status) else {
            let message = (try? JSONDecoder().decode([String: String].self, from: data))?["error"]
                ?? String(data: data, encoding: .utf8)
                ?? "unknown error"
            throw PhotoServiceError.server(status: status, message: message)
        }
        return try JSONDecoder().decode(T.self, from: data)
    }
}

File: mobile/src/services/photoService.ts
import { serverBaseURL } from '../config';
import { Coordinate } from '../geo';
import {
  AddFriendResponse,
  Explorer,
  FriendListResponse,
  RegistrationResponse,
  RemoteUser,
} from '../model/explorer';

export interface FriendStats {
  steps: number;
  exploredPercent: number;
}
import { NearbyResult, Photo, PhotoListResponse, UploadResponse } from '../model/photo';

/**
 * The shared half of Nimbus: leaving photos at a place and finding the ones
 * other people left.
 *
 * Kept as an interface so the local Node server can be swapped for a hosted
 * backend (Supabase, Firebase, anything) without touching a single view.
 */
export interface PhotoService {
  health(): Promise<boolean>;

  /**
   * Announce this device's identity and get back its friend code and current
   * friends. Safe to call on every launch; it doubles as a rename.
   */
  register(explorer: Explorer, stats?: FriendStats): Promise<RegistrationResponse>;
  friends(viewerID: string): Promise<RemoteUser[]>;
  addFriend(code: string, viewerID: string): Promise<AddFriendResponse>;

  nearby(coordinate: Coordinate, viewerID: string): Promise<NearbyResult>;
  photos(bounds: BoundingBox, viewerID: string): Promise<Photo[]>;
  upload(request: UploadRequest): Promise<UploadResponse>;

  /** Remove every photo this explorer has left. Never touches anyone else's. */
  deleteMyPhotos(viewerID: string): Promise<void>;
}

export interface BoundingBox {
  minLat: number;
  maxLat: number;
  minLon: number;
  maxLon: number;
}

export interface UploadRequest {
  imageBase64: string;
  coordinate: Coordinate;
  caption: string;
  explorer: Explorer;
  placeName: string | null;
}

export class PhotoServiceError extends Error {
  readonly status: number | null;
  /**
   * What the server itself said, without the status code wrapped round it.
   * "no one has that code" is the whole of what someone typing a friend code
   * needs to read; the status belongs in a log.
   */
  readonly serverMessage: string | null;

  constructor(message: string, status: number | null = null, serverMessage: string | null = null) {
    super(message);
    this.name = 'PhotoServiceError';
    this.status = status;
    this.serverMessage = serverMessage;
  }

  static server(status: number, message: string): PhotoServiceError {
    return new PhotoServiceError(`Server said ${status}: ${message}`, status, message);
  }

  static unreachable(): PhotoServiceError {
    return new PhotoServiceError("Can't reach the Nimbus server. Is `node server.js` running?");
  }
}

/** How long a request may take before it counts as unreachable. */
const REQUEST_TIMEOUT_MS = 8_000;
const UPLOAD_TIMEOUT_MS = 20_000;

/** Talks to server/server.js. */
export class NimbusAPI implements PhotoService {
  private readonly baseURL: string;

  constructor(baseURL: string = serverBaseURL) {
    this.baseURL = baseURL.replace(/\/$/, '');
  }

  async health(): Promise<boolean> {
    try {
      const response = await this.fetch('/health', {}, REQUEST_TIMEOUT_MS);
      return response.ok;
    } catch {
      return false;
    }
  }

  register(explorer: Explorer, stats?: FriendStats): Promise<RegistrationResponse> {
    return this.post('/users', {
      id: explorer.id,
      displayName: explorer.displayName,
      color: explorer.colorHex,
      ...(stats == null ? {} : { steps: stats.steps, exploredPercent: stats.exploredPercent }),
    });
  }

  async friends(viewerID: string): Promise<RemoteUser[]> {
    const response = await this.get<FriendListResponse>('/friends', { userId: viewerID });
    return response.friends;
  }

  addFriend(code: string, viewerID: string): Promise<AddFriendResponse> {
    return this.post('/friends', { userId: viewerID, code });
  }

  nearby(coordinate: Coordinate, viewerID: string): Promise<NearbyResult> {
    return this.get('/photos/nearby', {
      lat: String(coordinate.latitude),
      lon: String(coordinate.longitude),
      viewerId: viewerID,
    });
  }

  async photos(bounds: BoundingBox, viewerID: string): Promise<Photo[]> {
    const response = await this.get<PhotoListResponse>('/photos/bbox', {
      minLat: String(bounds.minLat),
      maxLat: String(bounds.maxLat),
      minLon: String(bounds.minLon),
      maxLon: String(bounds.maxLon),
      // Without this the server would hand back everybody's photographs.
      viewerId: viewerID,
    });
    return response.photos;
  }

  upload({
    imageBase64,
    coordinate,
    caption,
    explorer,
    placeName,
  }: UploadRequest): Promise<UploadResponse> {
    return this.post(
      '/photos',
      {
        userId: explorer.id,
        displayName: explorer.displayName,
        color: explorer.colorHex,
        lat: coordinate.latitude,
        lon: coordinate.longitude,
        caption,
        takenAt: Date.now(),
        ...(placeName == null ? {} : { placeName }),
        imageBase64,
      },
      UPLOAD_TIMEOUT_MS
    );
  }

  deleteMyPhotos(viewerID: string): Promise<void> {
    return this.delete(`/photos?${new URLSearchParams({ userId: viewerID }).toString()}`);
  }

  // MARK: Transport

  private get<T>(path: string, query: Record<string, string>): Promise<T> {
    const search = new URLSearchParams(query).toString();
    return this.send<T>(`${path}${search ? `?${search}` : ''}`, {}, REQUEST_TIMEOUT_MS);
  }

  private delete<T>(path: string, timeoutMs = REQUEST_TIMEOUT_MS): Promise<T> {
    return this.send<T>(path, { method: 'DELETE' }, timeoutMs);
  }

  private post<T>(path: string, body: unknown, timeoutMs = REQUEST_TIMEOUT_MS): Promise<T> {
    return this.send<T>(
      path,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
      },
      timeoutMs
    );
  }

  private async send<T>(path: string, init: RequestInit, timeoutMs: number): Promise<T> {
    const response = await this.fetch(path, init, timeoutMs);
    const text = await response.text();

    if (!response.ok) {
      let message = text || 'unknown error';
      try {
        message = (JSON.parse(text) as { error?: string }).error ?? message;
      } catch {
        // Not JSON; the body itself is the best message available.
      }
      throw PhotoServiceError.server(response.status, message);
    }

    return JSON.parse(text) as T;
  }

  /** `fetch` with a deadline, so a dead server fails fast instead of hanging. */
  private async fetch(path: string, init: RequestInit, timeoutMs: number): Promise<Response> {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), timeoutMs);
    try {
      return await fetch(`${this.baseURL}${path}`, { ...init, signal: controller.signal });
    } catch {
      throw PhotoServiceError.unreachable();
    } finally {
      clearTimeout(timeout);
    }
  }
}
